engine/common/synchronization/engine.go

```diff
@@ -302,7 +303,15 @@ func (e *Engine) SubmitLocal(event interface{}) {
 func (e *Engine) Submit(channel network.Channel, originID flow.Identifier, event interface{}) {
 	err := e.process(originID, event)
 	if err != nil {
-		engine.LogError(e.log, err)
+		lg := e.log.With().
+			Err(err).
+			Str("channel", channel.String()).
+			Str("origin", originID.String()).
+			Logger()
+		if errors.Is(err, engine.IncompatibleInputTypeError) {
+			lg.Error().Msg("received message with incompatible type")
```

Comment:
@AlexHentschel Do we actually want to return after this line? If I understand correctly, this `if` statement captures the case where we get an incompatible input type from the network, right? In that case, we probably shouldn't crash? Without the `return` after this line we will proceed to the `lg.Fatal` which comes after.

Reply:
Good question. 

I think throwing fatal errors here is intended, as [explained here](https://github.com/onflow/flow-go/pull/1008) and [here](https://github.com/onflow/flow-go/pull/1001/files#diff-9876a3f5e3a13c1fe35fc34b77aa7cc88f3c970efe12f9f3abbf5d5d0ebce3c2R332)

>added new sentinel error engine.IncompatibleInputTypeError, which engines can raise, if they receive an input that does not have any of the expected types (previously, engines just errored with an unspecific error)
this allows us to reject inputs from the networking layer with incompatible types
In contrast, an engine should generally not receive an incompatible input from a trusted internal component within the node. This would likely be an implementation bug and should crash the node

Reply:
> In contrast, an engine should generally not receive an incompatible input from a trusted internal component within the node. This would likely be an implementation bug and should crash the node

I believe this second case is reflected in the code [here](https://github.com/onflow/flow-go/pull/1001/files#diff-9876a3f5e3a13c1fe35fc34b77aa7cc88f3c970efe12f9f3abbf5d5d0ebce3c2R295-R296). However, my concern about this piece of code is that it seems to represent the first case:

> error engine.IncompatibleInputTypeError, which engines can raise, if they receive an input that does not have any of the expected types (previously, engines just errored with an unspecific error)
this allows us to reject inputs from the networking layer with incompatible types

The code above is in `Submit` and not `SubmitLocal`, which means it is processing a message which came from the network, right? So if we receive an invalid input from the network layer, shouldn't we just log the error and continue? We only want to crash if the invalid input is from an internal component.

Reply:
thanks. Good catch. Yes, this would be from external source, where we could expect that they feed us with invalid inputs.  I intended to return here (hence logging the error), but I messed this up.

Reply:
Cool. Thanks for the catch @smnzhu